src/VideoRemise/VideoRemise/VideoChannel.cs

```diff
@@ -171,59 +196,198 @@ internal async Task ShutdownAsync()
-            mediaRecording = await currentCapture.PrepareLowLagRecordToStorageFileAsync(MediaEncodingProfile.CreateMp4(VideoEncodingQuality.Auto), file);
-            //mediaRecording = await currentCapture.PrepareLowLagRecordToStreamAsync(MediaEncodingProfile.CreateMp4(VideoEncodingQuality.Auto), currentRecordingStream);
+            mediaRecording = await currentCapture.PrepareLowLagRecordToStorageFileAsync(MediaEncodingProfile.CreateMp4(VideoEncodingQuality.Auto), currentFile);
             isRecording = true;
             return mediaRecording.StartAsync();
         }
 
+        internal async Task EndPhrase(int length, TriggerType triggerType)
+        {
+            if (!isRecording)
+            {
+                return;
+            }
+
+            await StopRecording();
+            PhraseRecording pr = new PhraseRecording(currentFile, length, triggerType);
+            actions.Add(pr);
+        }
+
         internal async Task<IAsyncAction> StopRecording()
         {
             isRecording = false;
             await mediaRecording.StopAsync();
             return mediaRecording.FinishAsync();
         }
 
-        internal void StartPlayback()
+        internal async void Trigger(int length, TriggerType triggerType)
         {
-            captureElement.Visibility = Visibility.Collapsed;
-            mediaPlayerElement.Visibility = Visibility.Visible;
-            showingLive = false;
-            mediaPlayerElement.Source = MediaSource.CreateFromStream(currentRecordingStream, MediaEncodingProfile.CreateMp4(VideoEncodingQuality.Auto).ToString());
-            mediaPlayerElement.MediaPlayer.Play();
+            if (!isRecording)
+            {
+                return;
+            }
+
+            await EndPhrase(length, triggerType);
+            await StartPhrase();
+            StartPlayback();
         }
 
-        internal async Task StartLoop(int length)
+        internal void PlayFromFile()
         {
-            if (isRecording)
-            {
-                await StopRecording();
-            }
             captureElement.Visibility = Visibility.Collapsed;
             mediaPlayerElement.Visibility = Visibility.Visible;
             showingLive = false;
-            activeSource = MediaSource.CreateFromStream(currentRecordingStream, MediaEncodingProfile.CreateMp4(VideoEncodingQuality.Auto).ToString());
-            mediaPlayerElement.Source = activeSource;
-            var duration = activeSource.Duration?.TotalSeconds ?? 0;
-            mediaPlayerElement.MediaPlayer.PlaybackSession.Position = System.TimeSpan.FromSeconds(Math.Max(duration - length, 0));
-            mediaPlayerElement.MediaPlayer.Play();
+            playing = true;
+
+            var (source, length) = actions.ElementAt(currentReplay).GetSourceAndLength();
+            mediaPlayerElement.Source = source;
+            mediaPlayerElement.MediaPlayer.MediaOpened += (MediaPlayer sender, object args) =>
```

Comment:
Isn't there a possible timing issue if you do this after setting the source? Shouldn't this line and the previous be in the opposite order? Or is it always safe?

Reply:
I guess technically there could be a timing issue, but given the time scales involved I would expect it to be quite rare.

Reply:
Fixed